File: src/data/courses/dsa/linked-lists/singly-ll-core/deletion-variations/code/java/DeletionVariations.java
// ============================================================
// Deletion Variations
// ============================================================
class Solution {
    public ListNode deleteHead(ListNode head) { return head != null ? head.next : null; }
    public ListNode deleteTail(ListNode head) {
        if (head==null||head.next==null) return null;
        ListNode c = head; while (c.next.next!=null) c=c.next; c.next=null; return head;
    }
    public ListNode deleteValue(ListNode head, int val) {
        ListNode dummy = new ListNode(0); dummy.next = head; ListNode prev = dummy;
        while (prev.next!=null) { if (prev.next.val==val) { prev.next=prev.next.next; break; } prev=prev.next; }
        return dummy.next;
    }
}

File: src/data/courses/dsa/linked-lists/singly-ll-core/deletion-variations/code/python/deletion_variations.py
# ============================================================
# Deletion Variations
# ============================================================
class Solution:
    def deleteHead(self, head): return head.next if head else None
    def deleteTail(self, head):
        if not head or not head.next: return None
        curr = head
        while curr.next.next: curr = curr.next
        curr.next = None; return head
    def deleteValue(self, head, val):
        dummy = ListNode(0); dummy.next = head; prev = dummy
        while prev.next:
            if prev.next.val == val: prev.next = prev.next.next; break
            prev = prev.next
        return dummy.next
